sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala

```diff
@@ -416,32 +416,31 @@ private[lf] object Speedy {
       * producing a text message.
       */
     private[speedy] override def handleException(excep: SValue.SAny): Control[Nothing] = {
-      @tailrec def unwind(): Option[KTryCatchHandler] =
+      @tailrec
+      def unwind(ptx: PartialTransaction): Option[KTryCatchHandler] =
         if (kontDepth() == 0) {
           None
         } else {
           popKont() match {
             case handler: KTryCatchHandler =>
-              ptx = ptx.rollbackTry
+              this.ptx = ptx.rollbackTry
               Some(handler)
             case _: KCloseExercise =>
-              ptx = ptx.abortExercises
-              unwind()
+              unwind(ptx.abortExercises)
             case k: KCheckChoiceGuard =>
               // We must abort, because the transaction has failed in a way that is
               // unrecoverable (it depends on the state of an input contract that
               // we may not have the authority to fetch).
-              clearKontStack()
-              clearEnv()
+              abort()
               k.abort()
             case KPreventException() =>
```

Comment:
As you noted online: we could also `abort()` here to prevent the accidental resuming of the machine.